### DISCUSSION TITLE
how to send multifile by requests module to FastAPI?

### DISCUSSION THREAD

**Discussion by jbc0708** [fastapi/fastapi#6416] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
[Server]
from fastapi import FastAPI,  File, UploadFile

app = FastAPI()

@app.post("/single")
async def check(file: bytes = File() ):
    print(len(file))
    return {"state": 200}

@app.post("/multi")
async def check(files: list[bytes] = File()  ):
    for file in files:
        print(len(file))
    return {"state": 200}

[Client]
import requests

files1 = {  "file": open("images/4_1_1.jpg", "rb") }
files2 = {
        "file1": open("images/4_1_1.jpg", "rb"),
        "file2": open("images/4_1_2.jpg", "rb")
}

r = requests.post("http://127.0.0.1:8000/single", files=files1)
result = r.json()
print(result["state"])

r = requests.post("http://127.0.0.1:8000/multi", files=files2)
result = r.json()
print(result["state"])

I have problem about transfer multi files from client to Server..
```

### Description

Single file send dont have issue..
but multi file send always have issue..

result of single file send:
Server:     INFO:     127.0.0.1:54755 - "POST /single HTTP/1.1" 200 OK
Client:    200

result of multi file send:
Server:     INFO:     127.0.0.1:54761 - "POST /multi HTTP/1.1" 422 Unprocessable Entity
Client:    KeyError: "state"

How can i fix it..? please help me

### Operating System

macOS

### FastAPI Version

0.78.0

### Python Version

3.10.4

**Comment by arunppsg** [comment]
From the request files [docs](https://fastapi.tiangolo.com/tutorial/request-files/), the multifile upload argument should probably be like this:
```
from typing import List

@app.post("/multi")
async def check(files: List[UploadFile]):
    for file in files:
        print(len(file))
    return {"state": 200}
```

**Comment by jbc0708** [comment]
> From the request files [docs](https://fastapi.tiangolo.com/tutorial/request-files/), the multifile upload argument should probably be like this:
> 
> ```
> from typing import List
> 
> @app.post("/multi")
> async def check(files: List[UploadFile]):
>     for file in files:
>         print(len(file))
>     return {"state": 200}
> ```

Your code for pytho3.6 and above.. right?
But I use python3.10 ...
and I know that instant of UploadFile can't use len()  method...

**Comment by JarroVGIT** [comment]
The main reason you have a problem is because you are sending with two different names. Multiple files in FastAPI are received as a `List` of either `bytes` or `UploadFile` objects. I would recommend the latter, as the first is pure in-memory and the latter actually gives you greater control and allows for larger files. As the parameter in your path operation function is called `files`, it can't handle the post you are making (mimicking form-input with names `file1` and `file2`, used as keys in your dictionary when making the request). 

I have created a working sample for you here: https://github.com/JarroVGIT/fastapi-github-issues/tree/master/4992

Note that I use `UploadFile` in my FastAPI app. In my client, I give a list of tuples like so:
`(form-input-name, (filename, filehandler, file_info))`

The app:
```python
from fastapi import FastAPI,  File, UploadFile
import uvicorn

app = FastAPI()

@app.post("/multi")
async def check_multi_files(files: list[UploadFile]):
    return {"filenames": [file.filename for file in files]}

if __name__ == "__main__":
    uvicorn.run(app, host="0.0.0.0", port=8000)
```

The client:
```python
import requests

url = 'http://localhost:8000/multi'
multiple_files = [('files', ('file1.txt', open('files/file1.txt', 'rb'), 'text/plain')),
                      ('files', ('file2.txt', open('files/file2.txt', 'rb'), 'text/plain'))]
r = requests.post(url, files=multiple_files)
print(r.text)
```

Hope this clarifies.

**Reply by mkmkmkmkmkmkmkmkmkmkmk** [reply]
1. My front-end code sends the request code like this
![image](https://github.com/tiangolo/fastapi/assets/90512746/3f77127d-fc63-490b-b7d4-4b354c2eff5d)
2. The backend code looks like this：
![image](https://github.com/tiangolo/fastapi/assets/90512746/e5c3ea0f-6e00-4cdd-8814-a42ced5186c8)
3. console.log(fd.getAll("file"))输出如下：
![image](https://github.com/tiangolo/fastapi/assets/90512746/98837e30-80e2-4d7a-8cd6-13e294f2cdfe)
4.Why does the 422 error occur when accessing this interface '/v1/uploadFiles/'? How should I receive this fd request data on the backend? How should I modify it？
![image](https://github.com/tiangolo/fastapi/assets/90512746/3902f32b-670d-4697-8717-8f46ba895676)

（vesion: fastapi： 0.109.0 python:3.11.5）

**Comment by jbc0708** [comment, marked as answer]
> ```python
> @app.post("/multi")
> async def check_multi_files(files: list[UploadFile]):
>     return {"filenames": [file.filename for file in files]}
> ```
Hi JarroVGIT~!. First of all, Thank you.
I have coded as per your suggestion and succeeded in uploading multi-files.
Thank you~!

**Comment by bengabp** [comment]
I want to create a route to upload multiple files where each file has a name and description field like saving a list of identified documents for a user, how will i write my fastapi route to handle all files and get their names and descriptions correctly;

in a json payload it could look like this

```json
{
	"files":[
		{"name": "Residential document": "description":"This is a residential document", "file": "some bytes data"},
		{"name": "Residential document": "description":"This is a residential document", "file": "some bytes data"},
		{"name": "Residential document": "description":"This is a residential document", "file": "some bytes data"},

	]
}
```

I know i can create one route to this this for a single file , but it would mean that the client would need to call my api route multiple times in a row. Is this a good design pattern or is there a better way

**Reply by iudeen** [reply]
This doesn't seem like standard multipart file upload request. 

You can always create your own request with normal post request to meet your needs.
